go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -70,14 +68,127 @@ type NetworkControllerManager struct {
 	// used for leader election
 	identity string
 
-	// controller for all networks, key is netName of net-attach-def, value is *Controller
-	// this map is updated either at the very beginning of ovnkube-master when initializing the default controller
-	// or when net-attach-def is added/deleted. All these are serialized and no lock protection is needed
-	ovnControllers map[string]NetworkController
+	defaultNetworkController BaseNetworkController
+
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *netAttachDefinitionController
+}
+
+func (cm *networkControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	topotype := netConfInfo.GetTopologyType()
+	if netConfInfo.GetTopologyType() == ovntypes.Layer3AttachDefTopoType {
+		return ovn.NewSecondaryLayer3NetworkController(cnci, nInfo, netConfInfo), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topotype)
+}
+
+func (cm *networkControllerManager) NewDummyNetworkController(topoType, netName string) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	netInfo := util.NewNetInfo(&ovncnitypes.NetConf{NetConf: types.NetConf{Name: netName}, IsSecondary: true})
+	switch topoType {
+	case ovntypes.Layer3AttachDefTopoType:
+		return ovn.NewSecondaryLayer3NetworkController(cnci, netInfo, &util.Layer3NetConfInfo{}), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
+}
+
+// Find all the OVN logical switches/routers for the secondary networks
+func findAllSecondaryNetworkLogicalEntities(nbClient libovsdbclient.Client) ([]*nbdb.LogicalSwitch,
+	[]*nbdb.LogicalRouter, error) {
+	p1 := func(item *nbdb.LogicalSwitch) bool {
+		_, ok := item.ExternalIDs[ovntypes.NetworkNameExternalID]
+		return ok
+	}
+	nodeSwitches, err := libovsdbops.FindLogicalSwitchesWithPredicate(nbClient, p1)
+	if err != nil {
+		klog.Errorf("Failed to get all logical switches of secondary network error: %v", err)
+		return nil, nil, err
+	}
+	p2 := func(item *nbdb.LogicalRouter) bool {
+		_, ok := item.ExternalIDs[ovntypes.NetworkNameExternalID]
+		return ok
+	}
+	clusterRouters, err := libovsdbops.FindLogicalRoutersWithPredicate(nbClient, p2)
+	if err != nil {
+		klog.Errorf("Failed to get all distributed logical routers: %v", err)
+		return nil, nil, err
+	}
+	return nodeSwitches, clusterRouters, nil
+}
+
+func (cm *networkControllerManager) SyncNetworkControllers(allControllers []NetworkController) error {
+	existingNetworksMap := map[string]bool{}
+	for _, oc := range allControllers {
+		existingNetworksMap[oc.GetNetworkName()] = true
+	}
+
+	// Get all the existing secondary networks and its logical entities
+	switches, routers, err := findAllSecondaryNetworkLogicalEntities(cm.nbClient)
+	if err != nil {
+		return err
+	}
+
+	var ops []ovsdb.Operation
+	staleNetworks := map[string]NetworkController{}
```

Comment:
can we rename this variable ? I was confused in the caller when I saw a controller being put in a map named `staleNetworks` .